File: model/src/main/scala/com/lightbend/modelServer/model/ModelToServe.scala
package com.lightbend.modelServer.model

import java.io._

import com.lightbend.model.modeldescriptor.ModelDescriptor
import com.lightbend.model.winerecord.WineRecord

import scala.collection.Map
import com.lightbend.modelServer.model.PMML.PMMLModel
import com.lightbend.modelServer.model.tensorflow.TensorFlowModel

import scala.util.Try

/**
 * Created by boris on 5/8/17.
 */
object ModelToServe {
  private val factories = Map(
    ModelDescriptor.ModelType.PMML.name -> PMMLModel,
    ModelDescriptor.ModelType.TENSORFLOW.name -> TensorFlowModel
  )

  private val factoriesInt = Map(
    ModelDescriptor.ModelType.PMML.index -> PMMLModel,
    ModelDescriptor.ModelType.TENSORFLOW.index -> TensorFlowModel
  )

  def fromByteArray(message: Array[Byte]): Try[ModelToServe] = Try {
    val m = ModelDescriptor.parseFrom(message)
    m.messageContent.isData match {
      case true => new ModelToServe(m.name, m.description, m.modeltype, m.getData.toByteArray, m.dataType)
      case _ => throw new Exception("Location based is not yet supported")
    }
  }

  def fromModelToServe(descriptor : ModelToServe): Try[Model] = Try{
    println(s"New model - $descriptor")
    factories.get(descriptor.modelType.name) match {
      case Some(factory) => factory.create(descriptor)
      case _ => throw new Throwable("Undefined model type")
    }
  }

  def factoryOrdinal(ordinal : Int) : Option[ModelFactory] = factoriesInt.get(ordinal)
}

case class ModelToServe(name: String, description: String,
  modelType: ModelDescriptor.ModelType, model: Array[Byte], dataType: String) {}

case class ModelWithData(model: Option[ModelToServe], data : Option[WineRecord])

File: model/src/main/java/com/lightbend/model/ModelToServe.java
package com.lightbend.model;

import com.lightbend.model.PMML.PMMLModelFactory;
import com.lightbend.model.tensorflow.TensorflowModelFactory;

import java.io.Serializable;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

// Intermediate model representation used for transporting models
public class ModelToServe implements Serializable {

    private static final Map<Integer, ModelFactory> factories = new HashMap<Integer, ModelFactory>() {
        {
            put(Modeldescriptor.ModelDescriptor.ModelType.TENSORFLOW.getNumber(), TensorflowModelFactory.getInstance());
            put(Modeldescriptor.ModelDescriptor.ModelType.PMML.getNumber(), PMMLModelFactory.getInstance());
        }
    };

    private String name;
    private String description;
    private Modeldescriptor.ModelDescriptor.ModelType modelType;
    private byte[] modelData;
    private String modelDataLocation;
    private String dataType;

    public ModelToServe(String name, String description, Modeldescriptor.ModelDescriptor.ModelType modelType,
                        byte[] dataContent, String modelDataLocation, String dataType){
        this.name = name;
        this.description = description;
        this.modelType = modelType;
        this.modelData = dataContent;
        this.modelDataLocation = modelDataLocation;
        this.dataType = dataType;
    }

    public String getName() {
        return name;
    }

    public String getDescription() {
        return description;
    }

    public Modeldescriptor.ModelDescriptor.ModelType getModelType() {
        return modelType;
    }

    public String getDataType() {
        return dataType;
    }

    public byte[] getModelData() {
        return modelData;
    }

    public String getModelDataLocation() {
        return modelDataLocation;
    }

    @Override
    public String toString() {
        return "ModelToServe{" +
                "name='" + name + '\'' +
                ", description='" + description + '\'' +
                ", modelType=" + modelType +
                ", dataType='" + dataType + '\'' +
                '}';
    }

    public static Model restore(int type, byte[] bytes){
        ModelFactory factory = factories.get(type);
        if (factory == null) {
            System.out.println("Unknown model type " + type);
            return null;
        }
        return factory.restore(bytes);
    }

    public static Optional<Model> create(ModelToServe descriptor){
        ModelFactory factory = factories.get(descriptor.getModelType().ordinal());
        if (factory == null) {
            System.out.println("Unknown model type " + descriptor.getModelType());
            return Optional.empty();
        }
        return factory.create(descriptor);
    }
}
